File: OBSrange_MATLAB/functions/shootrays.m
function [rx,rz,Dr,rt,rv] = shootrays(p, v_profile, zmax, dr, vdz)
% [rx,rz,Dr,rt,rv] = shootrays(p, v_profile, zmax, dr, vdz)
% 
% 1D ray tracing function
% all units in km or km/s
% 
% Z. Eilon 01/2019 (modified from code by Brandon Schmandt)


if nargin < 4
    dr = 0.001;
end
if nargin<5
    vdz = 0.001;
end

zz1 = [0:vdz:(zmax-vdz)]';
zz2 = [vdz:vdz:zmax]';

v1  = linterp(v_profile.z,v_profile.ssp,zz1);
v2  = linterp(v_profile.z,v_profile.ssp,zz2);
% assume flat Earth (no radius terms)
u1 = 1./v1;
u2 = 1./v2; 

dv = v2-v1;

dz = zz2-zz1;
b = dv ./ dz;
const_indx = (b == 0);

X = zeros( length(v1), 1 );
X(const_indx) = constv_dist( u1(const_indx), dz(const_indx), p );
X(~const_indx) = gradv_dist(b(~const_indx),u1(~const_indx),u2(~const_indx),p);
X = [0; X];
X(imag(X)>0) = NaN;
X = X(~isnan(X));
Xd = cumsum(X);
rayz = [zz1(1); zz2]; 
rayz = rayz(1:length(X));
dz = dz(1:length(X)-1);

% calc Dr
Dr1 = (X(2:end).^2 + dz.^2).^(1/2);
if any(~isreal(Dr1)), error('Imaginary ray...'); end
Dr1 = [0; Dr1];
Dr2 = cumsum(Dr1);
Dr = unique([0:dr:max(Dr2),max(Dr2)]');
rx = interp1(Dr2,Xd,Dr);
rz = interp1(Dr2,rayz,Dr);

% calculate travel time (JBR)
dDr = Dr(2:length(Dr))-Dr(1:length(Dr)-1);
dDr = [0; dDr];
v = [v1(1); v2];
rv = interp1(Dr2,v,Dr);
dt = dDr./rv;
rt = cumsum(dt);
end

%-------------------- sub-functions --------------------------------------

function x = constv_dist( u, dz, p )
    x = (p * dz) ./ eta(u,p);
end

function x = gradv_dist( b, u1, u2, p )
    x1 = eta(u1,p) ./ (b.*u1*p);
    x2 = eta(u2,p) ./ (b.*u2*p);
    x = x1 - x2;
end

function val = eta( u, p )
    val = sqrt( u.^2 - p.^2 );
end

File: OBSrange_PYTHON/funcs/shootrays.py
'''
FUNCTION shootrays.py 

1D ray tracing function. All units in km or km/s

Z. Eilon 01/2019 (modified from code by Brandon Schmandt)
Translated to Python and further modified by S. Mosher 01/2019
'''
# Import modules and functions
import numpy as np
from numpy.lib.scimath import sqrt
from scipy.interpolate import interp1d

from IPython.core.debugger import Tracer

# Given two arrays, this function finds the closest value in the smaller array
# to each value in the larger array.
def find_ilay(r, R):
  N = len(r)
  ilay = np.zeros(N, dtype=int)

  for i, depth in enumerate(r):
    idx = np.where(R - depth > 0)[0][0]
    ilay[i] = idx

  return ilay

# Linear interpolation function.
def linterp(X, Y, XI):
  YI = np.zeros(len(XI))
  ilay = find_ilay(XI, X)
  
  # Linearly interpolate.
  YI = (XI - X[ilay-1])*(Y[ilay] - Y[ilay-1])/(X[ilay] - X[ilay-1]) + Y[ilay-1]

  # Sort out coincident elements.
  olap = np.intersect1d(X, XI)
  for i, val in enumerate(olap):
    YI[XI == olap[i]] = np.mean(Y[X == olap[i]])

  return YI

def shootrays(p, v_profile, zmax, dr=0.001, vdz=0.001):
  # Anonymous functions
  eta = lambda u, p: sqrt(u**2 - p**2)
  gradv_dist = lambda b, u1, u2, p: (eta(u1,p)/(b*u1*p)) - (eta(u2,p)/(b*u2*p))
  constv_dist = lambda u, dz, p: (p * dz) / eta(u, p)

  zz1 = np.arange(0, zmax, vdz)
  zz2 = np.arange(vdz, zmax + vdz, vdz)
  
  v1  = linterp(v_profile[0], v_profile[1], zz1)
  v2  = linterp(v_profile[0], v_profile[1], zz2)

  # Assume flat Earth (no radius terms)
  u1 = 1 / v1
  u2 = 1 / v2 
  
  dv = v2 - v1
  
  dz = zz2 - zz1
  b = dv / dz
  const_indx = (b == 0)

  X = np.zeros(len(v1), dtype=complex)
  
  X[const_indx == True] = \
                constv_dist( u1[const_indx == True], dz[const_indx == True], p)
  
  X[const_indx != True] = \
  gradv_dist(b[const_indx!=True], u1[const_indx!=True], u2[const_indx!=True], p)
  

  # When rays go complex skip out.
  if np.imag(X).any():
    rx = 0
    rz = 0
    Dr = 0
    rt = 0
    rv = 0

    return rx, rz, Dr, rt, rv

  else:
    X = np.real(X)
    X = np.hstack([0, X])
    Xd = np.cumsum(X)
    rayz = np.hstack([zz1[0], zz2]) 
    rayz = rayz[0:len(X)]
    dz = dz[0:len(X) - 1]
   
    # Calculate Dr
    Dr1 = np.sqrt(X[1:]**2 + dz**2)
    Dr1 = np.hstack([0, Dr1])
    Dr2 = np.cumsum(Dr1)
    Dr = np.unique( np.hstack([np.arange(0, max(Dr2), dr), max(Dr2)]))
    fx = interp1d(Dr2, Xd)
    fz = interp1d(Dr2, rayz)
    rx = fx(Dr)
    rz = fz(Dr)
  
    # Calculate travel-time (JBR)
    dDr = Dr[1:len(Dr)] - Dr[0:len(Dr)-1]
    dDr = np.hstack([0, dDr])
    v = np.hstack([v1[0], v2])
    fv = interp1d(Dr2, v)
    rv = fv(Dr)
    dt = dDr / rv
    rt = np.cumsum(dt)
  
    return rx, rz, Dr, rt, rv